### DISCUSSION TITLE
Can fragments define their own transitions?

### DISCUSSION THREAD

**Discussion by ovenum** [unpoly/unpoly#456] [Q&A] [answered]
Hello,
short question or suggestion regarding fragment transitions.

A link can update multiple fragments at once, but can only define one transition which is then applied to each fragment. 

Is it possible to define transitions by fragment selector, so that the fragment defines its transition and not the link? 
Listening to `up:fragment:inserted` is too late, since it does not allow to change the `renderOptions` anymore.

Listening to `up:fragmented:loaded` and checking `renderOptions.target` could work, but would require to call preventDefault() on the passed event and then calling `up.render()` for each render target with the new transition while setting history options on both renderOption objects passed to up.render().

```js
// Mockup code to illustrate the idea
up.on("up:fragment:loaded", (event) => {
  const { renderOptions, origin, request, response } = event;

  const targets = event.renderOptions.target.split(",").map(t => t.trim());
  
  // Only apply to custom targets  
  if (
    targets.includes(".target-selector-1") === false || 
    targets.includes(".target-selector-2") === false
  ) {
    return;
  }

  event.preventDefault();
 	
  up.render({
    target: ".target-selector-1",
    transition: "target-1-show",
    document: response.text,
    location: renderOptions.url,
    origin,
    navigate: true,
  });

  up.render({
    target: ".target-selector-2",
    transition: "target-2-show",
    document: response.text,
    origin,
    navigate: false,
  });
});
```

**Comment by triskweline** [comment, marked as answer]
The `{ transition }` option can be a function. Inside that function you can start multiple transitions by call `up.morph()` multiple times. Something like this (untested):

```js
up.on("up:fragment:loaded", ({ renderOptions }) => {
  renderOptions.transition = (oldElement, newElement) =>{
    let oldTarget1 = oldElement.querySelector('.target-selector-1')
    let newTarget1 = newElement.querySelector('.target-selector-1')
  
    if (oldTarget1 && newTarget1) {
      up.morph(oldTarget1, newTarget1, 'target-1-show')
    }

    let oldTarget2 = oldElement.querySelector('.target-selector-2')
    let newTarget2 = newElement.querySelector('.target-selector-2')
  
    if (oldTarget2 && newTarget2) {
      up.morph(oldTarget2, newTarget2, 'target-2-show')
    }
  }
})
```

For this to work the elements `.target-selector-1` and `.target-selector-2` must not be nested within each other.

**Reply by ovenum** [reply]
Thanks for looking into this and sketching a solution @triskweline! 

Using `up.morph` makes way more sense then calling `up.render` twice. 
I figured that `renderOptions.transition`, if passed a function, expects that function to fulfill the contract outlined in the [docs](https://unpoly.com/up.transition) for up.transition and by doing so, return a `Promise. In this case simply returning up.morph should be enough.

The v3 docs for [renderOptions.transition](https://v3.unpoly.com/up.render#options.transition) only list string as argument type. Maybe this could be updated or is it too early for this?

**Reply by ovenum** [reply]
Did not succeed in making it work. I have created a simplified [example on glitch.io ](https://glitch.com/edit/#!/rounded-groovy-trollius)(using v3.0.0-rc2) to illsutrate the issue better. Please note that overlay does not mean an unpoly.js overlay but rather another fragment placed on top of the pages main content. 

```js
up.on("up:fragment:loaded", ({ renderOptions }) => {
  renderOptions.transition = (oldElement, newElement) => {
    if (
      oldElement.hasAttribute("up-overlay") && 
      newElement.hasAttribute("up-overlay")
    ) {
      return up.morph(oldElement, newElement, "cross-fade", {
        duration: 2200
      });
    }
    
    return renderOptions.transition;
  };
});
```
Calling `up.morph()` will correctly create the transition DOM structure as described in motion.js. Not returning up.morph will skip the animation.
https://github.com/unpoly/unpoly/blob/19b3a8101c2d9244ac8b349eebee61a2a92b268a/src/unpoly/motion.js#L399-L404

But once the transition is finished the `oldElement` and the `newElement` element are removed from the DOM. 
Maybe i’m missing something here? Glad for any hints.

**Reply by triskweline** [reply]
I haven't had the chance for a closer look, but I notice that your `up.morph()` is using a longer duration instead of passing along the duration it got from `options`:

```js
up.on("up:fragment:loaded", ({ renderOptions }) => {
  renderOptions.transition = (oldElement, newElement, options) => {
    if (
      oldElement.hasAttribute("up-overlay") && 
      newElement.hasAttribute("up-overlay")
    ) {
      return up.morph(oldElement, newElement, "cross-fade", options);
    }
    
    return renderOptions.transition;
  };
});
```

Also note that prefixing your own custom attributes may clash with future Unpoly attributes.

**Reply by ovenum** [reply]
Thanks for spotting that mistake! Passing the options object to `up.morph()` will make it work.
Will drop the up- prefix for the overlay element.

**Reply by shybyte** [reply]
As far as I understand, the previous code does not work as intended. 
The line 
```js
return renderOptions.transition;
```
should return the original transition set via normal unpoly attributes.
But this does not work, it will return the modified transition.

Here is my solution, which keeps correctly the original transition and supports disabling a transition with "none" for certain elements:

```ts
up.on('up:fragment:loaded', ({ renderOptions }: { renderOptions: any }) => {
  const originalTransition = renderOptions.transition;

  renderOptions.transition = (oldElement: HTMLElement, newElement: HTMLElement, options: unknown) => {
    const transition = newElement.getAttribute('my-transition') || originalTransition;
    if (transition && transition !== 'none') {
      return up.morph(oldElement, newElement, transition, options);
    }
  };
});
```
